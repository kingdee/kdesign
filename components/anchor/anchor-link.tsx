import React, { useContext, useEffect, useRef } from 'react'
import classNames from 'classnames'
import AnchorContext from './context'
import ConfigContext from '../config-provider/ConfigContext'
import { usePrevious } from '../_utils/hooks'

export interface AnchorLinkProps {
  prefixCls?: string
  href: string
  target?: string
  title: React.ReactNode
  children?: React.ReactNode
  className?: string
  isFirstLevel?: boolean
}

const AnchorLink = (props: AnchorLinkProps): React.FunctionComponentElement<any> => {
  const { prefixCls: customPrefixcls, href, title, children, className, isFirstLevel } = props
  const { getPrefixCls, prefixCls } = useContext(ConfigContext)
  const anchorLinkPrefixCls = getPrefixCls!(prefixCls, 'anchor', customPrefixcls) // 锚点样式前缀

  const { activeLink, scrollTo, onClick, type, registerLink } = useContext(AnchorContext)

  const prevHref = usePrevious(href)
  const mounting = useRef(true)
  useEffect(() => {
    if (mounting.current) {
      registerLink(href)
      mounting.current = false
      return
    }
    if (prevHref !== href) {
      registerLink(href)
    }
  }, [href, registerLink, prevHref])

  const handleClick = (e: React.MouseEvent<HTMLElement>) => {
    onClick?.(e, { title, href })
    scrollTo(href)
  }

  const renderAnchorLink = () => {
    const active = activeLink === href
    const anchorLinkClasses = classNames(
      `${anchorLinkPrefixCls}-link`,
      {
        [`${anchorLinkPrefixCls}-link-active`]: active,
      },
      className,
    )

    const titleClassName = classNames(`${anchorLinkPrefixCls}-link-title`, {
      [`${anchorLinkPrefixCls}-link-title-active`]: active,
      [`${anchorLinkPrefixCls}-link-title-secondary`]: !isFirstLevel && type === 'bookmarks',
    })

    return (
      <div className={anchorLinkClasses}>
        <span title={typeof title === 'string' ? title : ''} className={titleClassName} onClick={handleClick}>
          {title}
        </span>
        {type !== 'menu' && children}
      </div>
    )
  }

  return renderAnchorLink()
}

export default AnchorLink
