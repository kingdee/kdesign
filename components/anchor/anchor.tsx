import React, { useState, useContext, useEffect, useCallback } from 'react'
import classNames from 'classnames'
import ConfigContext from '../config-provider/ConfigContext'
import { getCompProps } from '../_utils'
import { tuple } from '../_utils/type'
import AnchorContext from './context'
import devWarning from '../_utils/devwarning'
import { Icon } from '../index'
import usePopper from '../_utils/usePopper'
import { toArray } from '../_utils/react-children'
import { useResizeObserver } from '../_utils/hooks'

const sharpMatcherRegx = /#(\S+)$/

export type AnchorContainer = HTMLElement | Window

export const AnchorTypes = tuple('bookmarks', 'menu', 'advanced')
export type AnchorType = typeof AnchorTypes[number]

interface ScrollToOptions {
  getContainer?: () => HTMLElement | Window | Document
  callback?: () => any
  duration?: number
}

type Section = {
  link: string
  top: number
}

type anchorPositionProps = {
  width?: number
  height?: number
  top?: number
}

function getDefaultContainer() {
  return window
}

function getOffsetTop(element: HTMLElement, container: AnchorContainer): number {
  if (!element.getClientRects().length) {
    return 0
  }

  const rect = element.getBoundingClientRect()

  if (rect.width || rect.height) {
    if (container === window) {
      container = element.ownerDocument!.documentElement!
      return rect.top - container.clientTop
    }
    return rect.top - (container as HTMLElement).getBoundingClientRect().top
  }

  return rect.top
}

function isWindow(obj: any) {
  return obj !== null && obj !== undefined && obj === obj.window
}

function getScroll(target: HTMLElement | Window | Document | null, top: boolean): number {
  if (typeof window === 'undefined') {
    return 0
  }
  const method = top ? 'scrollTop' : 'scrollLeft'
  let result = 0
  if (isWindow(target)) {
    result = (target as Window)[top ? 'pageYOffset' : 'pageXOffset']
  } else if (target instanceof Document) {
    result = target.documentElement[method]
  } else if (target) {
    result = (target as HTMLElement)[method]
  }
  if (target && !isWindow(target) && typeof result !== 'number') {
    result = ((target as HTMLElement).ownerDocument || (target as Document)).documentElement?.[method]
  }
  return result
}

function easeInOutCubic(t: number, b: number, c: number, d: number) {
  const cc = c - b
  t /= d / 2
  if (t < 1) {
    return (cc / 2) * t * t * t + b
  }
  return (cc / 2) * ((t -= 2) * t * t + 2) + b
}

function scrollTo(y: number, options: ScrollToOptions = {}) {
  const { getContainer = () => window, callback, duration = 450 } = options
  const container = getContainer()
  const scrollTop = getScroll(container, true)
  const startTime = Date.now()

  const frameFunc = () => {
    const timestamp = Date.now()
    const time = timestamp - startTime
    const nextScrollTop = easeInOutCubic(time > duration ? duration : time, scrollTop, y, duration)
    if (isWindow(container)) {
      ;(container as Window).scrollTo(window.pageXOffset, nextScrollTop)
    } else if (container instanceof HTMLDocument || container.constructor.name === 'HTMLDocument') {
      ;(container as HTMLDocument).documentElement.scrollTop = nextScrollTop
    } else {
      ;(container as HTMLElement).scrollTop = nextScrollTop
    }
    if (time < duration) {
      window.requestAnimationFrame(frameFunc)
    } else if (typeof callback === 'function') {
      callback()
    }
  }
  window.requestAnimationFrame(frameFunc)
}

export interface AnchorProps {
  prefixCls?: string
  className?: string
  style?: React.CSSProperties
  children?: React.ReactNode
  type?: AnchorType
  offsetTop?: number
  bounds?: number
  affix?: boolean
  lockedIcon?: boolean | React.ReactNode
  icon?: React.ReactNode
  dropdownStyle?: React.CSSProperties
  visible?: boolean
  getContainer?: () => AnchorContainer
  getCurrentAnchor?: () => string
  onClick?: (e: React.MouseEvent<HTMLElement>, link: { title: React.ReactNode; href: string }) => void
  onChange?: (currentActiveLink: string) => void
}
const InternalAnchor = (props: AnchorProps, ref: unknown): React.FunctionComponentElement<AnchorProps> => {
  const { getPrefixCls, prefixCls, compDefaultProps: userDefaultProps } = useContext(ConfigContext)
  const anchorProps = getCompProps('Anchor', userDefaultProps, props)
  const {
    prefixCls: customPrefixcls,
    className,
    children,
    type,
    onChange,
    onClick,
    getCurrentAnchor,
    offsetTop = 0,
    targetOffset,
    getContainer,
    bounds = 5,
    style,
    lockedIcon = true,
    icon,
    affix,
    dropdownStyle,
    onVisibleChange,
  } = anchorProps

  const anchorPrefixCls = getPrefixCls!(prefixCls, 'anchor', customPrefixcls) // 锚点样式前缀

  devWarning(AnchorTypes.indexOf(type) === -1, 'anchor', `cannot found anchor type '${type}'`)

  const [activeLink, setActiveLink] = useState<string | null>(null)
  const [fixedTop, setFixedTop] = useState(false)
  const [links, setLinks] = useState<string[]>([])
  const [anchorPosition, setAnchorPosition] = useState<anchorPositionProps>({})
  const [listWidth, setListWidth] = useState(0)
  const [menuWrapWidth, setMenuWrapWidth] = useState(0)
  const [listPostion, setListPosition] = useState(0)
  const [positionHistory, setPositionHistory] = useState<number[]>([0])
  const [optionShow, setOptionShow] = useState<boolean>(!!props.visible)
  const [isLocked, setIsLocked] = useState(false)

  const isAdcanced = type === 'advanced'

  // ref
  const wrapperRef = React.useRef<HTMLDivElement>(null)
  const inkRef = React.useRef<HTMLElement>(null)
  const linkContentRef = React.useRef<HTMLDivElement>(null)
  const iconRef = React.useRef<HTMLElement>(null)
  const menuWrapRef = React.createRef<HTMLDivElement>()
  const normalRef = React.useRef<HTMLDivElement>(null)
  const anchorRef = (ref as any) || normalRef
  const linksWidthRef = React.useRef<number[]>([])

  const wrapperClass = classNames(className, {
    [`${anchorPrefixCls}-wrapper`]: true,
  })

  // classes
  const anchorMenuClass = classNames({
    [`${anchorPrefixCls}-menu`]: true,
  })

  const anchorAdvancedClass = classNames({
    [`${anchorPrefixCls}-advanced`]: true,
  })

  const inkClass = classNames(`${anchorPrefixCls}-line-slider`, {
    visible: activeLink,
  })

  React.useEffect(() => {
    setOptionShow(!!props.visible)
  }, [props.visible])

  const updateInk = useCallback(() => {
    const anchorNode = (isAdcanced ? wrapperRef : anchorRef).current
    const inkNode = inkRef.current
    const linkNode = anchorNode?.getElementsByClassName(`${anchorPrefixCls}-link-title-active`)[0]
    if (linkNode && inkNode) {
      inkNode.style.top = `${(linkNode as any).offsetTop + 1}px`
    }
  }, [anchorPrefixCls, anchorRef, isAdcanced])

  const mounting = React.useRef(true)
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    if (mounting.current) {
      mounting.current = false
      if (linkContentRef.current) {
        const nodelist = linkContentRef.current?.children
        const linksWidth = Array.from(nodelist).map((item: any) => item.getBoundingClientRect().width)
        linksWidthRef.current = linksWidth
      }
      return
    }
    updateInk()
  })

  useResizeObserver(anchorRef.current, () => {
    const nodelist = linkContentRef.current?.children
    if (!nodelist) return
    const linksWidth = Array.from(nodelist).map((item: any) => item.getBoundingClientRect().width)
    linksWidthRef.current = linksWidth
  })

  useEffect(() => {
    if (linkContentRef.current) {
      setTimeout(() => {
        const width = linkContentRef.current?.getBoundingClientRect().width || 0
        setListWidth(width)
      }, 0)
    }
  }, [linkContentRef, menuWrapRef, children])

  useEffect(() => {
    if (menuWrapRef.current) {
      setTimeout(() => {
        setMenuWrapWidth(menuWrapRef.current?.getBoundingClientRect().width || 0)
      }, 0)
    }
  }, [menuWrapRef])

  const registerLink = (link: string) => {
    if (!links.includes(link)) {
      setLinks((pre) => [...pre, link])
    }
  }

  const getScrollContainer = useCallback(() => {
    const getFunc = getContainer || getDefaultContainer
    return getFunc()
  }, [getContainer])

  const getAnchor = useCallback(
    (offsetTop = 0, bounds): string => {
      const linkSections: Array<Section> = []
      const container = getScrollContainer()
      links.forEach((link) => {
        const sharpLinkMatch = sharpMatcherRegx.exec(link.toString())
        if (!sharpLinkMatch) {
          return
        }
        const target = document.getElementById(sharpLinkMatch[1])
        if (target) {
          const top = getOffsetTop(target, container)
          if (top < offsetTop + bounds) {
            linkSections.push({
              link,
              top,
            })
          }
        }
      })
      if (linkSections.length) {
        const maxSection = linkSections.reduce((prev, curr) => (curr.top > prev.top ? curr : prev))
        return maxSection.link
      }
      return ''
    },
    [getScrollContainer, links],
  )

  const handleScrollTo = (link: string) => {
    setCurrentActiveLink(link)

    const container = getScrollContainer()
    const scrollTop = getScroll(container, true)
    const sharpLinkMatch = sharpMatcherRegx.exec(link)
    if (!sharpLinkMatch) {
      return
    }
    const targetElement = document.getElementById(sharpLinkMatch[1])
    if (!targetElement) {
      return
    }
    const eleOffsetTop = getOffsetTop(targetElement, container)
    let y = scrollTop + eleOffsetTop
    y -= targetOffset !== undefined ? targetOffset : offsetTop || 0
    scrollTo(y, {
      getContainer: getScrollContainer,
    })
  }

  const setCurrentActiveLink = useCallback(
    (link: string) => {
      if (activeLink === link) {
        return undefined
      }
      setActiveLink(typeof getCurrentAnchor === 'function' ? getCurrentAnchor() : link)
      onChange?.(link)
    },
    [onChange, setActiveLink, activeLink, getCurrentAnchor],
  )

  const handleScroll = useCallback(() => {
    if (affix && type !== 'advanced' && anchorRef.current) {
      const container = getScrollContainer()
      const top = getOffsetTop(anchorRef.current, container)
      if (top <= offsetTop) {
        setFixedTop(true)
      }
      if (top > offsetTop) {
        setFixedTop(false)
      }
    }
    const currentActiveLink = getAnchor(targetOffset !== undefined ? targetOffset : offsetTop || 0, bounds)
    currentActiveLink && setCurrentActiveLink(currentActiveLink)
  }, [
    affix,
    setFixedTop,
    offsetTop,
    bounds,
    getAnchor,
    setCurrentActiveLink,
    type,
    getScrollContainer,
    anchorRef,
    targetOffset,
  ])

  useEffect(() => {
    getScrollContainer().addEventListener('scroll', handleScroll)
    handleScroll()
    return () => getScrollContainer().removeEventListener('scroll', handleScroll)
  }, [handleScroll, getScrollContainer])

  useEffect(() => {
    if (!isAdcanced && anchorRef && anchorRef.current) {
      const { top, width, height } = anchorRef.current.getBoundingClientRect()
      setAnchorPosition({ top, width, height })
    }
  }, [anchorRef, getScrollContainer, isAdcanced])

  const handleVisibleChange = (visible: boolean) => {
    !isLocked && onVisibleChange && onVisibleChange(visible)
    if (props.visible !== undefined) return
    !isLocked && setOptionShow(visible)
  }

  const popperProps = {
    autoPlacement: false,
    ...anchorProps,
    prefixCls: anchorPrefixCls,
    defaultVisible: optionShow,
    visible: optionShow,
    gap: -(((ref as any) || iconRef).current?.offsetHeight || 0),
    onVisibleChange: handleVisibleChange,
  }

  const fixedRef = React.useRef<HTMLDivElement>(null)

  if (type === 'advanced' && isLocked) {
    if (fixedRef.current) {
      const { left, top } = fixedRef.current.getBoundingClientRect()
      popperProps.popperStyle = { position: 'fixed', left, top }
    }
  }

  const addChildrenProps = (linksChildren: any) => {
    return toArray(linksChildren).map((item: any, indx: number) => {
      if (!item) return item
      return React.cloneElement(item, { isFirstLevel: true, key: indx })
    })
  }

  const anchorBookmarksContent = (ref: any) => (
    <div ref={ref} className={wrapperClass} style={style}>
      <div
        className={anchorPrefixCls}
        style={fixedTop ? { position: 'fixed', top: offsetTop || 0, width: anchorPosition.width } : undefined}
      >
        <div className={`${anchorPrefixCls}-line`}>
          <span className={inkClass} ref={inkRef} />
        </div>
        {addChildrenProps(children)}
      </div>
    </div>
  )

  const renderLeftArrow = () => {
    const leftClasses = classNames(`${anchorMenuClass}-left-arrows`, {
      [`${anchorMenuClass}-left-arrows-disabled`]: listPostion === 0,
    })

    const handleLeft = () => {
      if (listPostion === 0) return
      const positions = [...positionHistory]
      positions.pop()
      setPositionHistory(positions)

      setListPosition(positions[positions.length - 1])
    }
    return (
      <span className={leftClasses} onClick={handleLeft}>
        <Icon type="arrow-left" />
      </span>
    )
  }

  const findCurrentLinkPosition = (linksWidth: number[], boxWidth: number, position: number) => {
    if (!linksWidth.length) return 0
    let offset = 0
    for (let index = 0; index < linksWidth.length - 1; index++) {
      if (linksWidth[index] + offset < boxWidth - position) {
        offset += linksWidth[index]
      }
    }
    return offset
  }

  const renderRightArrow = () => {
    const canActionRightArrow = listPostion === menuWrapWidth - listWidth
    const rightClasses = classNames(`${anchorMenuClass}-right-arrows`, {
      [`${anchorMenuClass}-right-arrows-disabled`]: canActionRightArrow,
    })
    const handleRight = () => {
      if (canActionRightArrow) return
      const offset = findCurrentLinkPosition(linksWidthRef.current, menuWrapWidth, listPostion)
      const posiostion = offset + menuWrapWidth >= listWidth ? menuWrapWidth - listWidth : -offset
      setPositionHistory((pre) => [...pre, posiostion])
      setListPosition(posiostion)
    }
    return (
      <span className={rightClasses} onClick={handleRight}>
        <Icon type="arrow-right" />
      </span>
    )
  }

  const anchorMenuContent = (
    <div ref={anchorRef} className={`${anchorPrefixCls}-menu-wrapper`} style={{ width: '100%', ...style }}>
      <div
        className={anchorMenuClass}
        style={
          fixedTop
            ? { position: 'fixed', top: offsetTop || 0, width: style?.width || '100px' }
            : { width: style?.width }
        }
      >
        {renderLeftArrow()}
        <div className={`${anchorPrefixCls}-menu-wrap`} ref={menuWrapRef} style={{ maxWidth: listWidth }}>
          <div
            ref={linkContentRef}
            className={`${anchorPrefixCls}-menu-link-list`}
            style={{ left: `${listPostion}px` }}
          >
            {children}
          </div>
        </div>
        {renderRightArrow()}
      </div>
    </div>
  )

  const renderAnchorAdvanced = () => {
    const iconList = toArray(lockedIcon)
    const unlock = iconList[0] || <Icon type="unlock-solid" />
    const lock = iconList[1] || <Icon type="lock-solid" />
    return (
      <div ref={fixedRef} className={anchorAdvancedClass} style={dropdownStyle}>
        {lockedIcon !== false && (
          <span className={`${anchorPrefixCls}-advanced-lock`} onClick={() => setIsLocked(!isLocked)}>
            {isLocked ? lock : unlock}
          </span>
        )}
        {anchorBookmarksContent(wrapperRef)}
      </div>
    )
  }

  const anchorAdvancedContent = usePopper(
    <span className={`${anchorPrefixCls}-advanced-arrows`} ref={isAdcanced ? (ref as any) || iconRef : null}>
      {icon || <Icon type="location-solid" />}
    </span>,
    renderAnchorAdvanced(),
    popperProps,
  )

  const renderAnchorContent = (anchorType: string) => {
    switch (anchorType) {
      case 'bookmarks':
        return anchorBookmarksContent(anchorRef)
      case 'menu':
        return anchorMenuContent
      case 'advanced':
        return anchorAdvancedContent
      default:
        return anchorBookmarksContent(anchorRef)
    }
  }

  return (
    <AnchorContext.Provider
      value={{
        registerLink,
        activeLink,
        scrollTo: handleScrollTo,
        onClick,
        type,
      }}
    >
      {renderAnchorContent(type)}
    </AnchorContext.Provider>
  )
}

const Anchor = React.forwardRef<unknown, AnchorProps>(InternalAnchor)
Anchor.displayName = 'Anchor'
export default Anchor
