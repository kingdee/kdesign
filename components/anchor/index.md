---
category: Components
type: 导航
order: 0
title: Anchor
subtitle: 锚点
---

锚点是一种用于跳转到页面指定位置的组件，分为 2 种类型:
- 纵向锚点
- 横向锚点

## 使用场景

在以下情况使用锚点
- 需要对较长内容进行快速定位时

如果出现以下情况，请勿使用锚点
- 需要查看的内容相对独立时，请改用[页签](/components/tabs/)

## API

### Anchor Props

| 属性 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- | --- |
| affix | 固定模式 | boolean | true | `true` `false` | 1.0.0 |
| bounds | 锚点区域边界 | number | 5 | - | 1.0.0
| dropdownStyle | 下拉面板的 style 属性(高级锚点有效) | CSSProperties | - |  | 1.0.0 |
| getContainer | 指定滚动的容器 | () => HTMLElement | () => window |  | 1.0.0 |
| getCurrentAnchor | 自定义高亮的锚点 | () => string | - |  | 1.0.0 |
| icon | 高级锚点显示图标 | ReactNode | - | - | 1.0.0
| lockedIcon | 自定义锁定图标：\[未锁定图标, 锁定后的图标]，为 false 时隐藏 | boolean \| \[ReactNode, ReactNode] | - |  | 1.0.0 |
| offsetTop | 距离窗口顶部指定偏移量后触发固定(affix为true时) | number |  |  | 1.0.0 |
| onChange | 监听锚点链接改变 | (currentActiveLink: string) => void | - |  | 1.0.0 |
| onClick | `click` 事件的 handler | function(e: Event, link: Object) | - |  | 1.0.0 |
| targetOffset | 锚点滚动偏移量，默认与 offsetTop 相同 | number | - | - | 1.5.1 |
| trigger | 触发行为(高级锚点有效)，可使用数组设置多个触发行为 | string \| string\[] | `hover` | `hover` \| `focus` \| `click` \| `contextMenu` | 1.0.0 |
| type | 锚点类型（类型为 `menu` 时只显示一级锚点） | string | `bookmarks` | `bookmarks` `menu` `advanced` | 1.0.0 |
| visible | 手动控制浮层显隐 | boolean | false | `true` `false` | 1.0.0 |

### Link Props

| 成员 | 说明 | 类型 | 默认值 | 可选值 | 版本 |
| --- | --- | --- | --- | --- |
| href | 锚点链接 | string | - |  | 1.0.0 |
| title | 文字内容 | ReactNode | - |  | 1.0.0 |

## Design Token

| 分类 | 组件token | 全局token | 默认值 |
| --- | --- | --- | --- |
| color | --kd-c-anchor-color-background | --kd-g-color-background | #fff |
|  | --kd-c-anchor-color-border | --kd-g-color-border-strong | #d9d9d9 |
|  | --kd-c-anchor-advanced-arrows-color-text | - | #666 |
|  | --kd-c-anchor-color-theme | --kd-g-color-theme | #5582f3 |
|  | --kd-c-anchor-color-border-disabled | --kd-g-color-border-disabled | #ccc |
|  | --kd-c-anchor-color-text-primary | --kd-g-color-text-primary | #212121 |
|  | --kd-c-anchor-color-text-secondary | --kd-g-color-text-secondary | #666 |
| font | --kd-c-anchor-advanced-lock-font-size | - | 16px |
|  | --kd-c-anchor-font-size | --kd-g-font-size-small | 12px |
|  | --kd-c-anchor-icon-font-size | - | 16px |
|  | --kd-c-anchor-advanced-arrows-font-size | - | 24px |
| motion | --kd-c-anchor-duration-duration | --kd-g-duration | 0.3s |
| sizing | --kd-c-anchor-sizing-width | - | 120px |
|  | --kd-c-anchor-advanced-sizing-width | - | 160px |
|  | --kd-c-anchor-horizontal-link-sizing-max-width | - | 96px |
| spacing | --kd-c-anchor-spacing-vertical | - | 7px |
|  | --kd-c-anchor-spacing-horizontal | - | 7px |
|  | --kd-c-anchor-spacing-padding | - | 7px 8px |
|  | --kd-c-anchor-secondary-spacing-padding | - | 12px 8px |
|  | --kd-c-anchor-advanced-spacing-padding-top | - | 8px |
|  | --kd-c-anchor-advanced-spacing-padding-bottom | - | 7px |
|  | --kd-c-anchor-line-slider-spacing-width | - | 1px |
