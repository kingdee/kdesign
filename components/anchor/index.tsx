import OriginAnchor, { AnchorProps } from './anchor'
import AnchorLink, { AnchorLinkProps } from './anchor-link'

export type OriginAnchorProps = typeof OriginAnchor & {
  Link: typeof AnchorLink
}
const Anchor = OriginAnchor as OriginAnchorProps
Anchor.Link = AnchorLink
export default Anchor
export { AnchorProps, AnchorLinkProps }
