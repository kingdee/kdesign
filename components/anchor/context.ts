import * as React from 'react'
import { AnchorProps } from './anchor'
export interface AnchorContextProps {
  registerLink?: (str: string) => void
  activeLink?: string | null
  scrollTo?: (str: string) => void
  onClick?: AnchorProps['onClick']
  type?: AnchorProps['type']
}
const AnchorContext = React.createContext<AnchorContextProps>({})

export default AnchorContext
