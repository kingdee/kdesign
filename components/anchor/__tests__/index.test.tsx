import React from 'react'
import { mount, render } from 'enzyme'
import Anchor from '../index'
import Icon from '../../icon'
import { AnchorTypes, AnchorType } from '../anchor'
import mountTest from '../../../tests/shared/mountTest'
import ConfigProvider from '../../config-provider/index'
describe('Anchor', () => {
  // 1.mount test
  describe('1.mount test', () => {
    mountTest(Anchor)
    AnchorTypes.forEach((type) => {
      mountTest(() => <Anchor type={type} />)
    })
  })
  // 2.render test
  describe('2.render test', () => {
    it('renders correctly', () => {
      expect(
        render(
          <Anchor>
            <Anchor.Link href="#使用场景" title="使用场景" />
          </Anchor>,
        ),
      ).toMatchSnapshot()
      AnchorTypes.forEach((type) => {
        expect(
          render(
            <Anchor type={type}>
              <Anchor.Link href="#使用场景" title="使用场景" />
            </Anchor>,
          ),
        ).toMatchSnapshot()
      })
    })
  })
  // 3.warns in component
  describe('3.warns in component', () => {
    it('warns if type is wrong', () => {
      const mockWarn = jest.fn()
      jest.spyOn(console, 'warn').mockImplementation(mockWarn)
      const type = 'who am I' as any as AnchorType
      render(<Anchor type={type} />)
      expect(mockWarn).toHaveBeenCalledTimes(1)
      expect(mockWarn.mock.calls[0][0]).toMatch("Warning: [kdesign]-anchor: cannot found anchor type 'who am I'")
    })
  })
  // 4. render null or undefined without errors
  describe('4.render null or undefined without errors', () => {
    it('render null or undefined buttons without errors', () => {
      expect(
        mount(
          <Anchor>
            {null}
            {undefined}
          </Anchor>,
        ),
      ).toMatchSnapshot()
    })
  })
  // 5. displayName
  describe('5.displayName', () => {
    it('should have displayName static property', () => {
      const wrapper = mount(<Anchor></Anchor>)
      expect((wrapper.type() as any).displayName).toBe('Anchor')
    })
  })
  // 6. class state
  describe('6.class state', () => {
    it('should class use right', () => {
      const DefaultAnchor = mount(
        <Anchor>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(DefaultAnchor.find(`.kd-anchor`).length).toBeGreaterThanOrEqual(1)

      const menuAnchor = mount(
        <Anchor type="menu">
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(menuAnchor.find(`.kd-anchor-menu`).length).toBeGreaterThanOrEqual(1)

      const advancedAnchor = mount(
        <Anchor type="advanced">
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(advancedAnchor.find(`.kd-anchor-advanced-arrows`).length).toBeGreaterThanOrEqual(1)
    })
  })

  // 7.component interaction(event)
  describe('7.component interaction(event)', () => {
    it('onClick', () => {
      const onClick = jest.fn()
      const wrapper = mount(
        <Anchor onClick={onClick}>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )

      wrapper.find('.kd-anchor-link-title').at(0).simulate('click')
      expect(onClick).toHaveBeenCalled()
    })

    it('onChange', () => {
      const onChange = jest.fn()
      const wrapper = mount(
        <Anchor onChange={onChange}>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )

      wrapper.find('.kd-anchor-link-title').at(0).simulate('click')
      expect(onChange).toHaveBeenCalled()
    })
  })

  // 8.config provider
  describe('8.config provider', () => {
    it('should config use config provider', () => {
      const anchorConfig = {
        compDefaultProps: {
          Anchor: {
            type: 'menu',
          },
        },
      }
      const wrapper = mount(
        <ConfigProvider value={anchorConfig}>
          <Anchor></Anchor>
        </ConfigProvider>,
      )
      expect(wrapper.find(`.kd-anchor-menu`).length).toBeGreaterThanOrEqual(1)
    })
  })

  // 9. ref test
  // 10. api test
  describe('10.api test', () => {
    it('dropdownStyle API test', () => {
      const test = mount(
        <Anchor type="advanced" visible dropdownStyle={{ color: 'red' }}>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(getComputedStyle(document.querySelector('.kd-anchor-advanced')!, null).getPropertyValue('color')).toBe(
        'red',
      )
      test.unmount()
    })
    it('icon API test', () => {
      // icon
      const iconDemo = mount(
        <Anchor type="advanced" icon={<Icon type="add"></Icon>}>
          <Anchor.Link href="#API" title="#API" />
        </Anchor>,
      )
      expect(iconDemo.exists('.kdicon-add')).toBeTruthy()
      // lockedIcon、visible
      const advanceAnchor = mount(
        <Anchor type="advanced" visible lockedIcon={false}>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(document.querySelector('.kd-anchor-advanced-lock')).toBeFalsy()
      advanceAnchor.unmount()
      const advanceAnchor2 = mount(
        <Anchor
          type="advanced"
          visible
          lockedIcon={[<Icon type="add" key={1}></Icon>, <Icon type="add" key={2}></Icon>]}
        >
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(document.querySelector('.kdicon-add')).toBeTruthy()
      advanceAnchor2.unmount()
    })
    it('lockedIcon and visible API test', () => {
      const advanceAnchor = mount(
        <Anchor type="advanced" visible lockedIcon={false}>
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(document.querySelector('.kd-anchor-advanced-lock')).toBeFalsy()
      advanceAnchor.unmount()
      const advanceAnchor2 = mount(
        <Anchor
          type="advanced"
          visible
          lockedIcon={[<Icon type="add" key={1}></Icon>, <Icon type="add" key={2}></Icon>]}
        >
          <Anchor.Link href="#使用场景" title="使用场景" />
        </Anchor>,
      )
      expect(document.querySelector('.kdicon-add')).toBeTruthy()
      advanceAnchor2.unmount()
    })
  })
})
