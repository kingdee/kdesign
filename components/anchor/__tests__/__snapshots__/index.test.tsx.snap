// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor 2.render test renders correctly 1`] = `
<div
  class="kd-anchor-wrapper"
>
  <div
    class="kd-anchor"
  >
    <div
      class="kd-anchor-line"
    >
      <span
        class="kd-anchor-line-slider"
      />
    </div>
    <div
      class="kd-anchor-link"
    >
      <span
        class="kd-anchor-link-title"
        title="使用场景"
      >
        使用场景
      </span>
    </div>
  </div>
</div>
`;

exports[`Anchor 2.render test renders correctly 2`] = `
<div
  class="kd-anchor-wrapper"
>
  <div
    class="kd-anchor"
  >
    <div
      class="kd-anchor-line"
    >
      <span
        class="kd-anchor-line-slider"
      />
    </div>
    <div
      class="kd-anchor-link"
    >
      <span
        class="kd-anchor-link-title"
        title="使用场景"
      >
        使用场景
      </span>
    </div>
  </div>
</div>
`;

exports[`Anchor 2.render test renders correctly 3`] = `
<div
  class="kd-anchor-menu-wrapper"
  style="width:100%"
>
  <div
    class="kd-anchor-menu"
  >
    <span
      class="kd-anchor-menu-left-arrows kd-anchor-menu-left-arrows-disabled"
    >
      <i
        class="kdicon kdicon-arrow-left"
      />
    </span>
    <div
      class="kd-anchor-menu-wrap"
      style="max-width:0"
    >
      <div
        class="kd-anchor-menu-link-list"
        style="left:0px"
      >
        <div
          class="kd-anchor-link"
        >
          <span
            class="kd-anchor-link-title"
            title="使用场景"
          >
            使用场景
          </span>
        </div>
      </div>
    </div>
    <span
      class="kd-anchor-menu-right-arrows kd-anchor-menu-right-arrows-disabled"
    >
      <i
        class="kdicon kdicon-arrow-right"
      />
    </span>
  </div>
</div>
`;

exports[`Anchor 2.render test renders correctly 4`] = `
<span
  class="kd-anchor-advanced-arrows"
>
  <i
    class="kdicon kdicon-location-solid"
  />
</span>
`;

exports[`Anchor 4.render null or undefined without errors render null or undefined buttons without errors 1`] = `
<Anchor>
  <div
    className="kd-anchor-wrapper"
  >
    <div
      className="kd-anchor"
      style={
        Object {
          "position": "fixed",
          "top": 0,
          "width": 0,
        }
      }
    >
      <div
        className="kd-anchor-line"
      >
        <span
          className="kd-anchor-line-slider"
        />
      </div>
    </div>
  </div>
</Anchor>
`;
